src/nmodl/deriv.cpp

```diff
@@ -657,17 +666,19 @@ if (_deriv%d_advance) {\n",
         q = insertsym(q4, sp);
         eqnqueue(q);
         Sprintf(buf,
-                "_p[_dlist%d[_id]] - (_p[_slist%d[_id]] - _savstate%d[_id])/d%s;\n",
+                "_ml->data(_iml, _dlist%d[_id]) - (_ml->data(_iml, _slist%d[_id]) - "
+                "_savstate%d[_id])/d%s;\n",
                 numlist,
                 numlist,
                 numlist,
                 indepsym->name);
         Insertstr(q4, buf);
-        Sprintf(buf,
-                "}else{\n_dlist%d[++_counte] = _p[_slist%d[_id]] - _savstate%d[_id];}}}\n",
-                numlist + 1,
-                numlist,
-                numlist);
+        Sprintf(
+            buf,
+            "}else{\n_dlist%d[++_counte] = _ml->data(_iml, _slist%d[_id]) - _savstate%d[_id];}}}\n",
```

Comment:
In principle the function call can get inlined, yes.

The actual data access is a little more complicated:
- unlike CoreNEURON, there is not a single data array. this means there is a different base pointer for each RANGE variable
  - a contiguous array of these base pointers is maintained, so there is some `double** base_ptrs` somewhere
  - a contiguous array of the array sizes of the (array) RANGE variables is also maintained (`int* array_sizes`)
- because of the SoAoS format for array RANGE variables, the offset is not just `_iml`
  - it has to be calculated from the index into the array RANGE variable, the size of the array RANGE variable, and `_iml`
- that boils down to something like `base_ptrs[idx.field_index][array_sizes[idx.field_index] * _iml + idx.array_index]` where `idx` is the second argument, `_slist%d[_id]`

In practice in the generated code then you know all the sizes/indices at compile time, and some of them are normally zeros and ones, so we can make that more explicit if need be (i.e. if this turns out to be a bottleneck). But if the loop over `_iml` is visible (which it usually is for the "heavy" kernels) then the compiler should be able to tell that the stride/offset/base pointer are all invariants.